### PR TITLE
Always include trailing newline

### PR DESCRIPTION
**Pull request by camdencheek**
The goal of this PR is to fix our "line model" to fix the edge cases that led to https://github.com/sourcegraph/sourcegraph/issues/60605. In short, this changes the definition of a "line" to include its terminating newline (if it exists).

Before this PR, we had defined a "line" as starting at the byte after a newline (or the beginning of a file) and ending at the byte before a newline (or the end of the file).

The problem with that definition is that a newline that is the last byte in the file _can never successfully be matched_ because we would trim that from the returned content, so any ranges that would match that trailing newline would be out of bounds in the result returned to the client. That's the reason behind the panics caused by https://github.com/sourcegraph/zoekt/pull/709, which was an attempt to formalize the "line does not include a trailing newline" definition. 

So, instead, this PR proposes that we redefine a line as ending at the byte _after_ a newline (or the end of the file). This means that a regex can successfully and safely match a terminating newline. 

The downside is it does complicate the contract for the client a bit. In practice, it means to get the set of lines, you need to do something like `chunk.content.replace(/\r?\n$/, '').split(/\r?\n/)` instead of just `chunk.content.split(/\r?\n/)` because there may or may not be a trailing newline at the end of the file, but if there is, it does not indicate there is an empty line at the end of the file.

This PR makes significant client-observable changes, and would be accompanied by some associated updates in sourcegraph/sourcegraph drafted [here](https://github.com/sourcegraph/sourcegraph/pull/61247).

### CODE REVIEW

File: contentprovider.go [hunk 1]
```diff
@@ -451,41 +458,52 @@ type newlines struct {
...
-	return idx + 1, int(start), int(end)
+	// TODO: make sure this makes sense not including the newline
+	return idx + 1
 }
 
-// lineBounds returns the byte offsets of the start and end of the 1-based
-// lineNumber. The end offset is exclusive and will not contain the line-ending
-// newline. If the line number is out of range of the lines in the file, start
-// and end will be clamped to [0,fileSize].
-func (nls newlines) lineBounds(lineNumber int) (start, end uint32) {
```
Comment (camdencheek) [inline contentprovider.go:491]: I split `lineBounds` into `lineStart` and `lineEnd` because 1) there were often places where we only needed one or the other, and 2) it let me add an option to exclude the trailing newline for `lineEnd` more easily.

File: contentprovider.go [hunk 2]
```diff
@@ -495,20 +513,8 @@ func (nls newlines) getLines(data []byte, low, high int) []byte {
...
-	// the start of a new line, so we should not either. lineBounds clamps to
-	// len(data) when an out-of-bounds line is requested.
-	//
-	// As an example, if we request lines 1-5 from a file with contents
-	// `one\ntwo\nthree\n`, we should return `one\ntwo\nthree` because those are
-	// the three "lines" in the file, separated by newlines.
-	if highEnd == uint32(len(data)) && bytes.HasSuffix(data, []byte{'\n'}) {
-		highEnd = highEnd - 1
-		lowStart = min(lowStart, highEnd)
-	}
```
Comment (camdencheek) [inline contentprovider.go:535]: This is the culprit of the panics

File: contentprovider.go [hunk 3]
```diff
@@ -451,41 +458,52 @@ type newlines struct {
...
-		end = 0
+		return 0
 	} else if endIdx >= len(nls.locs) {
-		end = nls.fileSize
+		return nls.fileSize
+	} else if trimNewline {
+		return nls.locs[endIdx]
 	} else {
-		end = nls.locs[endIdx]
+		return nls.locs[endIdx] + 1
```
Comment (camdencheek) [inline contentprovider.go:505]: Primary change: if `trimNewline == false`, we include the newline.

**Comment by camdencheek** [comment]
@sourcegraph/search-platform mind taking a peek at this to see if y'all agree with the direction?

**Comment by isker** [comment]
https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/V1_chap03.html#tag_03_206 History is on your side 🌞.

**Comment by keegancsmith** [comment]
> In short, this changes the definition of a "line" in a chunk to include its terminating newline (if it exists).

This sounds good and correct to me.

> So, instead, this PR proposes that we redefine a line as ending at the byte _after_ a newline (or the end of the file). This means that a regex can successfully and safely match a terminating newline.

This makes a lot of sense to me. If your regex asks for `\n` and we say sure we found one but then it isn't there... that feels broken. So I 100% agree we need to go in this direction.

> The downside is it does complicate the ChunkMatch contract for the client a bit. In practice, it means to get the set of lines, you need to do something like `chunk.content.replace(/\r?\n$/, '').split(/\r?\n/)` instead of just `chunk.content.split(/\r?\n/)` because there may or may not be a trailing newline at the end of the file, but if there is, it does not indicate there is an empty line at the end of the file.

It is already a bit tricky for the client right? It feels like we should document corner cases for content for clients, to ensure they test. Additionally what you propose I think will match what users want to see.

Some fun corner cases while I am here. Imagine your corpus is two files: `` (empty) and `\n` (single newline):

- `/^\n$/` 
- `/^\n?$/` 

> This PR makes client-observable changes, and would be accompanied by some associated updates in sourcegraph/sourcegraph drafted [here](https://github.com/sourcegraph/sourcegraph/pull/61247).

I'm happy with the description. Ready for code review?

**Comment by camdencheek** [comment]
@keegancsmith, this is ready for review now. It doesn't look we do any "release" for Zoekt -- any thoughts about how to publicize this since it is a breaking change?

File: contentprovider.go [hunk 4]
```diff
@@ -316,19 +318,24 @@ func (p *contentProvider) fillContentChunkMatches(ms []*candidateMatch, numConte
...
 			startOffset := cm.byteOffset
 			endOffset := cm.byteOffset + cm.byteMatchSz
-			startLine, startLineOffset, _ := newlines.atOffset(startOffset)
-			endLine, endLineOffset, _ := newlines.atOffset(endOffset)
+			startLine := newlines.atOffset(startOffset)
+			endLine := newlines.atOffset(
+				// We want the line of the last byte in the match, not the first byte outside of the match.
+				// For a zero-length match, endOffset-1 could be before match start, so fall back to the
+				// byte after the match (as we do for startLine), not before.
+				max(startOffset, endOffset-1),
```
Comment (keegancsmith) [inline contentprovider.go]: you can't just - 1, endOffset is an unsigned int at could be 0.

File: contentprovider.go [hunk 5]
```diff
@@ -219,15 +219,17 @@ func (p *contentProvider) fillContentMatches(ms []*candidateMatch, numContextLin
...
+		lineEnd := int(p.newlines().lineEnd(num))
 
 		var lineCands []*candidateMatch
 
 		endMatch := m.byteOffset + m.byteMatchSz
 
 		for len(ms) > 0 {
 			m := ms[0]
-			if int(m.byteOffset) <= lineEnd {
+			if int(m.byteOffset) < lineEnd {
```
Comment (keegancsmith) [inline contentprovider.go]: is lineEnd the position of the newline, or newline + 1?

File: contentprovider.go [hunk 6]
```diff
@@ -219,15 +219,17 @@ func (p *contentProvider) fillContentMatches(ms []*candidateMatch, numContextLin
 	var result []LineMatch
 	for len(ms) > 0 {
 		m := ms[0]
-		num, lineStart, lineEnd := p.newlines().atOffset(m.byteOffset)
+		num := p.newlines().atOffset(m.byteOffset)
+		lineStart := int(p.newlines().lineStart(num))
+		lineEnd := int(p.newlines().lineEnd(num))
```
Comment (keegancsmith) [inline contentprovider.go]: now that you make lineEnd point to +1, shouldn't you just get rid of the function and instead ask here `lineEnd = p.newlines().lineStart(num + 1)`. Then I would also consider a different variable name, like nextLine. Reason I say this is I keep reviewing this code and wondering if things should be exclusive/inclusive/etc. My mental model has an easier time if I just know the offset is the start of the next line (which I believe this is now?)

File: contentprovider.go [hunk 7]
```diff
@@ -250,7 +252,7 @@ func (p *contentProvider) fillContentMatches(ms []*candidateMatch, numContextLin
 		// crosses a line boundary. Prevent confusion by
 		// taking lines until we pass the last match
 		for lineEnd < len(data) && endMatch > uint32(lineEnd) {
```
Comment (keegancsmith) [inline contentprovider.go]: I'm a bit confused why endMatch > lineEnd comparison doesn't change, but I can't work out why it was wrong before. Now this seems correct... I think.

File: contentprovider.go [hunk 8]
```diff
@@ -316,19 +318,24 @@ func (p *contentProvider) fillContentChunkMatches(ms []*candidateMatch, numConte
...
+				// byte after the match (as we do for startLine), not before.
+				max(startOffset, endOffset-1),
+			)
 
 			ranges = append(ranges, Range{
 				Start: Location{
 					ByteOffset: startOffset,
 					LineNumber: uint32(startLine),
-					Column:     columnHelper.get(startLineOffset, startOffset),
+					Column:     columnHelper.get(int(newlines.lineStart(startLine)), startOffset),
```
Comment (keegancsmith) [inline contentprovider.go:351]: may be very minor perf, but can you calculate these both just once like we did before.

File: internal/e2e/e2e_rank_test.go [hunk 9]
```diff
@@ -311,7 +311,7 @@ func marshalMatches(w io.Writer, rq rankingQuery, q query.Q, files []zoekt.FileM
 		chunks, hidden := splitAtIndex(f.ChunkMatches, chunkMatchesPerFile)
 
 		for _, m := range chunks {
-			_, _ = fmt.Fprintf(w, "%d:%s%s\n", m.ContentStart.LineNumber, string(m.Content), addTabIfNonEmpty(m.DebugScore))
+			_, _ = fmt.Fprintf(w, "%d:%s%s", m.ContentStart.LineNumber, string(m.Content), addTabIfNonEmpty(m.DebugScore))
```
Comment (keegancsmith) [inline internal/e2e/e2e_rank_test.go]: could you rather trim the trailing newline here? Otherwise when we enable debug scoring it will print very weirdly

**Review by keegancsmith — CHANGES_REQUESTED** [review]
I still need to review this deeper, so requesting changes so I remember to come back to it + I have some blocking feedback.

**Comment by keegancsmith** [comment]
> @keegancsmith, this is ready for review now. It doesn't look we do any "release" for Zoekt -- any thoughts about how to publicize this since it is a breaking change?

Yeah we don't have a changelog. So I am just going to CC people who I know use our API. cc @isker for neogrok and @binarymason from GitLab.

Alternatively is it possible to only do this trimming behaviour for ChunkMatches? Or is that asking for trouble.

**Comment by camdencheek** [comment]
>  is it possible to only do this trimming behaviour for ChunkMatches?

I did try that in my first attempt. Not impossible, but definitely messy. Also, the ambiguity described exists for `LineMatch` too, so I consider this a bugfix. If you'd like, I can make a sibling PR that attempts this and we can decide whether it's worth further splitting the code paths.

File: contentprovider.go (hunk 5 above)
Comment (camdencheek) [reply contentprovider.go]: Newline + 1. The intent is `data[lineStart:lineEnd]` is the full line. Previously, we did not include the newline character in the, so we used `<=` so we don't add a newline-matching range to the following line. So this was already treating a matched newline as "part of the line" here

File: contentprovider.go (hunk 6 above)
Comment (camdencheek) [reply contentprovider.go]: Good call. I hadn't even thought of that. Agreed that really simplifies the mental model.

File: contentprovider.go (hunk 4 above)
Comment (camdencheek) [reply contentprovider.go]: Oof, good catch, thanks. This is one of those moments where I could really use a nice [saturating_sub](https://doc.rust-lang.org/std/intrinsics/fn.saturating_sub.html). 

Can you think of a better API her that would make zero-length matches less awkward? I guess the problem is that the "boundary between two bytes" could be interpreted as the end of the first byte or the beginning of the second byte.

File: internal/e2e/e2e_rank_test.go (hunk 9 above)
Comment (camdencheek) [reply internal/e2e/e2e_rank_test.go]: Yeah, can do.

File: contentprovider.go (hunk 6 above)
Comment (camdencheek) [reply contentprovider.go]: Done in https://github.com/sourcegraph/zoekt/pull/747/commits/f2d9b8f0e3ce8c28ac4220c44a73c472c695df16

File: contentprovider.go (hunk 8 above)
Comment (camdencheek) [reply contentprovider.go:351]: Not sure I understand 🤔 I don't see any duplicate calculations?

File: internal/e2e/e2e_rank_test.go (hunk 9 above)
Comment (camdencheek) [reply internal/e2e/e2e_rank_test.go]: Done in https://github.com/sourcegraph/zoekt/pull/747/commits/1f9b00fb1d410f439bb4c2b76133fcfd0e2bc99c

File: contentprovider.go (hunk 4 above)
Comment (camdencheek) [reply contentprovider.go]: Fixed in https://github.com/sourcegraph/zoekt/pull/747/commits/2200f349eb1a01fe4bdf4ac580475f7b78a08ef3

File: contentprovider.go (hunk 7 above)
Comment (camdencheek) [reply contentprovider.go]: Yeah, I was also confused by this. The conclusion I came to was that, even though we considered a newline byte past the end of the line, if we matched that newline, we had to fetch all the way to the end of the following line so that the content included the matched newline and the matched range did not overflow the content.

**Comment by binarymason** [comment]
Thanks for the ping @keegancsmith! The logic behind this PR makes sense to me, but I'll need to dig in a bit to see if there would be any breaking changes on our end. We pin to a specific commit, so I don't see any blockers that should prevent this from moving forward.

If I see any breaking changes, will report back ASAP. 🤝

File: contentprovider.go (hunk 8 above)
Comment (keegancsmith) [reply contentprovider.go:351]: yeah sorry thought this was in a loop.

File: contentprovider.go [hunk 10]
```diff
@@ -316,19 +318,24 @@ func (p *contentProvider) fillContentChunkMatches(ms []*candidateMatch, numConte
...
 			startOffset := cm.byteOffset
 			endOffset := cm.byteOffset + cm.byteMatchSz
-			startLine, startLineOffset, _ := newlines.atOffset(startOffset)
-			endLine, endLineOffset, _ := newlines.atOffset(endOffset)
+			startLine := newlines.atOffset(startOffset)
+			endLine := newlines.atOffset(
+				// We want the line of the last byte in the match, not the first byte outside of the match.
+				// For a zero-length match, endOffset-1 could be before match start, so fall back to the
+				// byte after the match (as we do for startLine), not before.
+				max(startOffset, max(endOffset, 1)-1 /* prevent underflow */),
```
Comment (keegancsmith) [inline contentprovider.go]: you do this in one other place where you need to protect for underflow. Maybe it makes sense for a little helper like
```
startLine, endLine := newlines.atOffsetInclusiveRange(startOffset, endOffset)
```

File: contentprovider.go [hunk 11]
```diff
@@ -375,8 +382,8 @@ func chunkCandidates(ms []*candidateMatch, newlines newlines, numContextLines in
 	for _, m := range ms {
 		startOffset := m.byteOffset
 		endOffset := m.byteOffset + m.byteMatchSz
-		firstLine, _, _ := newlines.atOffset(startOffset)
-		lastLine, _, _ := newlines.atOffset(endOffset)
+		firstLine := newlines.atOffset(startOffset)
+		lastLine := newlines.atOffset(max(startOffset, endOffset-1))
```
Comment (keegancsmith) [inline contentprovider.go]: here is the other place with the underflow.

File: index_test.go [hunk 12]
```diff
@@ -202,7 +202,7 @@ func (s *memSeeker) Size() (uint32, error) {
 func TestNewlines(t *testing.T) {
 	b := testIndexBuilder(t, nil,
 		Document{Name: "filename", Content: []byte("line1\nline2\nbla")})
-	// ---------------------------------------------012345-678901-234
+	// -------------------------------------------012345-678901-234
```
Comment (keegancsmith) [inline index_test.go]: not sure if just github display, but this doesn't align before or after. I would suspect the 0 needs to line up with the very first char of the string.

File: index_test.go [hunk 13]
```diff
@@ -3254,7 +3254,7 @@ func TestLineAnd(t *testing.T) {
 
 	t.Run("ChunkMatches", func(t *testing.T) {
 		res := searchForTest(t, b, &q, chunkOpts)
-		wantRegexpCount := 1
+		wantRegexpCount := 2 // TODO: justify this change
```
Comment (keegancsmith) [inline index_test.go]: justification?

**Review by keegancsmith — APPROVED** [review]
LGTM!

Can you add docstrings to api.go for the LineEnd on LineMatches (and I guess for ChunkMatch related fields as well).

File: index_test.go (hunk 12 above)
Comment (camdencheek) [reply index_test.go]: Oh yep. My editor's tabwidth made made this not line up because the comment didn't start at the same indent level. Moved the comment above the line which should make it look right for everyone

File: index_test.go (hunk 13 above)
Comment (camdencheek) [reply index_test.go]: Took a while to find it (first time I've ever had two debuggers going in parallel), but I had broken a perf optimization. I missed a spot where I needed to change `<= to `<`. Glad I left the TODO, and thanks for reminding me!

**Comment by isker** [comment]
This was a [breaking change](https://github.com/isker/neogrok/commit/4ef496127fa60cc8030e61c53a07902360ec990f#diff-facc61c2995a11cde43af8ce3de520ee67b74a238167af846b2f32d82fc0d65f) for neogrok, but one that I'm happy to handle.

**Comment by keegancsmith** [comment]
Sorry and thank you @isker!
